gateway/hail.nginx.conf.in

```diff
@@ -158,3 +161,113 @@ server {
     include /etc/letsencrypt/options-ssl-nginx.conf;
     ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
 }
+
+server {
+    server_name app.@domain@;
+
+    gzip on;
+    gzip_types  text/plain text/css text/html text/javascript application/javascript application/x-javascript text/js text/xml  application/xml application/json application/rss+xml image/svg+xml font/opentype;
+
+    location / {
+        proxy_pass http://web/;
+    }
+
+    listen [::]:443 ssl;
+    listen 443 ssl;
+    ssl_certificate /etc/letsencrypt/fullchain.pem;
+    ssl_certificate_key /etc/letsencrypt/privkey.pem;
+    include /etc/letsencrypt/options-ssl-nginx.conf;
+    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
+}
+
+server {
+    server_name notebook-api.@domain@;
+
+    # Support large URI's stemming from access tokens
+    client_header_buffer_size 64k;
+    large_client_header_buffers 4 64k;
```

Comment:
We had a brief discussion the other day about passing the token when redirect to notebook.  Would making the token a cookie solve this?  Is that better that the header?

Reply:
Yeah, in my response to your question about /auth-notebook I mention this as a potential strategy. This is maybe the preferred way for the 2nd ... Nth access of /instance/<svc_name>.

The initial request needs to either have an access_token get parameter, or, if we're ok with not having a portable url, we can have app.hail.is do something more complicated, where the listed notebook isn't simply an anchor tag `<a href'=https://notebook-api.hail.is.instance/svc-...` but some click handler that sets a cookie and then redirects the user.

Reply:
I found that response way too specific to the details of Jupyter (discussion of file types, etc.)

Won't making the JWT a cookie instead of putting it in the header completely solve this problem?  Then all accesses (1st and Nth) check the user in the JWT and are the same, you don't need a second cookie, we get a portable URL and the link to the notebook is a regular anchor, etc.  No?

Reply:
If the JWT is a cookie, one of our services needs to set it. That happens in one of two scenarios: app.hail.is sets it when it receives the token from auth0 after authentication, or Nginx gateway sets it after auth_request reads/authorizes an ?access_token ... 

You of course can't pin a cookie to a URL, since the URL is fully defined by its string domain/path/to/resource?query_variables . So when I say "portable" I mean that the url will only work from a browser that already has the cookie set.

Reply:
That's correct.

Reply:
To elaborate, if we set a domain for the cookie, all subdomains of that domain will have access to it. However, the nginx gateway will not see that cookie if the /instance/... url is accessed from a browser in which that cookie wasn't already sent.

So if we want the instance url to be copy-pastable this won't work, only having the access_token in the url on the first request will work.

We probably prefer to require a cookie, since the cookie will represent one added assurance (that the url was accessed from app.hail.is), although we should not feel too confident in this as an access control, because the access token can, and over enough time and use, will, be leaked. In that case, the only access control is the access token's short lifespan.